### PR TITLE
Translate SimpleCase expression to connector expression

### PR DESCRIPTION
**Pull request by guiyanakuang**
## Description
Add translation for SimpleCaseExpression to connector expression form.
> Is this change a fix, improvement, new feature, refactoring, or other?

New feature.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine & postgres connector.

> How would you describe this change to a non-technical end user or system administrator?

This enabled pushdown of SimpleCase which are not expressible with Domain (i.e. references other symbols or contains function calls)

## Additional context and related issues
`GenericRewrite` cannot support translation of variable-length functions.
This pr adds several classes such as `VariableLengthCallPattern` to pattern match the entire argument list. Instead of matching each parameter individually (as variable length functions cannot do this).

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# PostgreSQL connector
* Improve performance of queries involving `SimpleCase` expression by pushing predicate computation to the PostgreSQL database. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by guiyanakuang** [comment]
I've sent the signed CLA. 😄

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 1]
```diff
@@ -651,6 +654,49 @@ protected Optional<ConnectorExpression> visitCast(Cast node, Void context)
             return Optional.empty();
         }
 
+        @Override
+        protected Optional<ConnectorExpression> visitSimpleCaseExpression(SimpleCaseExpression node, Void context)
```
Comment (assaf2) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:658]: Please also add a translation from `Call` back to `SimpleCaseExpression` at `ConnectorToSqlExpressionTranslator#translateCall`

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 2]
```diff
@@ -651,6 +654,49 @@ protected Optional<ConnectorExpression> visitCast(Cast node, Void context)
...
+                Optional<ConnectorExpression> defaultValue = process(node.getDefaultValue().get());
+                if (defaultValue.isEmpty()) {
+                    return Optional.empty();
+                }
+                arguments.add(defaultValue.get());
+            }
+
+            return Optional.of(new Call(typeOf(node), SIMPLE_CASE_WHEN_FUNCTION_NAME, arguments.build()));
+        }
+
```
Comment (assaf2) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:699]: Please add a test for this translation to `TestConnectorExpressionTranslator`

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java (hunk 1 above)
Comment (guiyanakuang) [reply core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:705]: Addressed comments. PTAL @assaf2

**Comment by guiyanakuang** [comment]
Would greatly appreciate if one of you could review this pr. Thank you for your time and efforts.

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 3]
```diff
@@ -498,6 +505,49 @@ protected Optional<Expression> translateInPredicate(ConnectorExpression value, C
...
+                return Optional.empty();
+            }
+
+            Optional<Expression> operand = translate(arguments.get(0));
+            if (operand.isEmpty()) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<WhenClause> builder = ImmutableList.builder();
+            for (int i = 1; i < argumentsSize - 1; i += 2) {
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:521]: This seems complex. Can we model it in the following way instead:

`$simple_case_when(operand, $array($simple_case_when_condition(condition, value)), [defaultValue])`

@findepi thoughts?

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 4]
```diff
@@ -498,6 +505,49 @@ protected Optional<Expression> translateInPredicate(ConnectorExpression value, C
...
+                Optional<WhenClause> whenClause = translateWhenClause(arguments.get(i), arguments.get(i + 1));
+                if (whenClause.isPresent()) {
+                    builder.add(whenClause.get());
+                }
+                else {
+                    return Optional.empty();
+                }
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:528]: You can flip condition here to return earlier:
```suggestion
               if (whenClause.isEmpty()) {
           	      return Optional.empty();
               }
                
               builder.add(whenClause.get());
```

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java [hunk 5]
```diff
@@ -498,6 +505,49 @@ protected Optional<Expression> translateInPredicate(ConnectorExpression value, C
...
+
+        protected Optional<WhenClause> translateWhenClause(ConnectorExpression operandValue, ConnectorExpression resultValue)
+        {
+            Optional<Expression> operand = translate(operandValue);
+            Optional<Expression> result = translate(resultValue);
+            if (operand.isPresent() && result.isPresent()) {
+                return Optional.of(new WhenClause(operand.get(), result.get()));
+            }
+            else {
+                return Optional.empty();
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:547]: `else` is redundant as you are returning above

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java [hunk 6]
```diff
@@ -492,6 +496,50 @@ public void testTranslateIn()
                 Optional.empty());
     }
 
+    @Test
+    public void testTranslateSimpleCase()
+    {
+        double doubleValue = 50.0d;
+        String strValue1 = "abc";
```
Comment (wendigo) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:503]: do not abbreviate: just `stringValue1` or `value1`

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java [hunk 7]
```diff
@@ -492,6 +496,50 @@ public void testTranslateIn()
...
+                new WhenClause(new DoubleLiteral(String.valueOf(doubleValue)), new StringLiteral(strValue1)));
+        Expression defaultValue = new StringLiteral(strValue2);
+
+        ConnectorExpression operandConnectorExpression = new Variable("double_symbol_1", DOUBLE);
+
+        List<ConnectorExpression> whenClauseConnectorExpressions = ImmutableList.of(
+                new Variable("double_symbol_2", DOUBLE),
+                new Variable("varchar_symbol_1", VARCHAR_TYPE),
+                new Constant(doubleValue, DOUBLE),
+                new Constant(Slices.wrappedBuffer(strValue1.getBytes(UTF_8)), createVarcharType(strValue1.length())));
```
Comment (wendigo) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:517]: static import `wrappedBuffer`

File: core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java [hunk 8]
```diff
@@ -492,6 +496,50 @@ public void testTranslateIn()
...
+
+        ConnectorExpression defaultValueConnectorExpression =
```
Comment (wendigo) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestConnectorExpressionTranslator.java:519]: just `defaultValue` (we already know that this variable is of `ConnectorExpression` type so no need to repeat it)

File: lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/expression/ConnectorExpressionIndex.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+{
+    private final Type type;
+
+    private final int index;
+
+    private final boolean isLast;
+
+    public ConnectorExpressionIndex(Type type, int index, boolean isLast)
+    {
+        this.type = type;
```
Comment (wendigo) [inline lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/expression/ConnectorExpressionIndex.java:28]: `requireNonNull(type, "type is null")`

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/ExpressionListCapture.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    private final String name;
+
+    private final Capture<List<? extends ConnectorExpression>> capture = newCapture();
+    private final Pattern<List<? extends ConnectorExpression>> pattern;
+
+    public ExpressionListCapture(String name, Predicate<ConnectorExpressionIndex> predicate)
+    {
+        this.name = requireNonNull(name, "name is null");
+        this.pattern = Pattern.typeOfList(ConnectorExpression.class)
+                .matching(expressionTypes(predicate))
```
Comment (wendigo) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/ExpressionListCapture.java:42]: requireNonNull(predicate, "predicate is null")

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+import static com.google.common.collect.MoreCollectors.toOptional;
+import static io.trino.spi.expression.StandardFunctions.SIMPLE_CASE_WHEN_FUNCTION_NAME;
+import static java.lang.String.format;
+
+public class RewriteSimpleCaseExpression
+        implements ConnectorExpressionRule<Call, String>
+{
+    public static RewriteSimpleCaseExpression createRewrite()
+    {
+        return createRewrite(Collections.emptyMap(), Optional.empty(), Optional.empty());
```
Comment (wendigo) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java:39]: `ImmutableMap.of()`

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                            rewritten.append(" WHEN ").append(rewrittenExpression.get());
+                        }
+                        else if (i % 2 == 0) {
+                            rewritten.append(" THEN ").append(rewrittenExpression.get());
+                        }
+                        else {
+                            rewritten.append(" ELSE ").append(rewrittenExpression.get());
+                        }
+                    }
+                    else {
```
Comment (wendigo) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java:128]: if you flip condition here and throw first, there will be less nesting

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        MatchContext matchContext = new MatchContext();
+        variableLengthCallPattern.resolve(captures, matchContext);
+
+        StringBuilder rewritten = new StringBuilder();
+
+        rewritten.append("(CASE ");
+
+        Optional<Object> capture = matchContext.getIfPresent("arguments");
+
+        if (capture.isPresent()) {
```
Comment (wendigo) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/expression/RewriteSimpleCaseExpression.java:105]: flip condition and exit early

**Review by wendigo — CHANGES_REQUESTED** [review]
Just skimmed

**Comment by guiyanakuang** [comment]
Thank you for your review. It's already late at night where I am, so I will need to address these issues tomorrow.

**Comment by wendigo** [comment]
Sure thing @guiyanakuang

File: core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java (hunk 3 above)
Comment (guiyanakuang) [reply core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java:521]: > This seems complex. Can we model it in the following way instead:
> 
> `$simple_case_when(operand, $array($simple_case_when_condition(condition, value)), [defaultValue])`
> 
> @findepi thoughts?

This is a great idea, and I will do some exploration.

**Comment by guiyanakuang** [comment]
> `$simple_case_when(operand, $array($simple_case_when_condition(condition, value)), [defaultValue])`

Based on the suggested model, I have refactored the code. Also solved all code style issues.

@wendigo PTAL : )

**Comment by guiyanakuang** [comment]
@findepi, could you help review this PR and let me know if there is anything that needs further improvement? Thank you

**Comment by findepi** [comment]
@martint had some thoughts on how the CASE should be modeled in connector expressions

**Comment by mosabua** [comment]
:wave: @guiyanakuang @martint  - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

**Pull request closed without merging**